app/models/event.rb

```diff
@@ -23,5 +23,27 @@ class Event < ApplicationRecord
 
   belongs_to :handleable, polymorphic: true, optional: true
   belongs_to :project
-  validates :name, :data, presence: true
+
+  validates :name, presence: true
+  validates :data, presence: true
+  validates :handleable, presence: true, if: proc { |event| event.handled_type? }
+  validate :name_matches_handleable_type, if: proc { |event| handleable && event.handled_type? }
+
+  scope :received_after, ->(time) { time ? where('created_at > ?', time) : all }
+
+  ##
+  # Validate that the receiver name matches the class of the handleable Event.
+  # For example if handleable is a Events::PullRequest object the name is
+  # expected to be 'pull_request'
+  def name_matches_handleable_type
```

Comment:
I'd use a custom validator to prevent as much as possible having code in models https://guides.rubyonrails.org/active_record_validations.html#custom-validators

Reply:
@santib  will do. Thanks!